File: CPP/Lesson_123_design_hashset.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 123 -- Design HashSet
// Category   : Arrays and Hashing
// Difficulty : Easy
// Study Plan : Day 62
// =============================================================
//
// QUESTION:
//   Design a HashSet (without built-in set): add, remove, contains.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct MyHashSet{ vector<vector<int>> b=vector<vector<int>>(769); int h(int k){return k%769;} void add(int k){auto& x=b[h(k)];if(find(x.begin(),x.end(),k)==x.end())x.push_back(k);} void remove(int k){auto& x=b[h(k)];x.erase(std::remove(x.begin(),x.end(),k),x.end());} bool contains(int k){auto& x=b[h(k)];return find(x.begin(),x.end(),k)!=x.end();}};
int main(){MyHashSet s; s.add(1); s.add(2); cout<<boolalpha<<s.contains(1)<<" "<<s.contains(3)<<"\n"; s.remove(2); cout<<s.contains(2)<<"\n";}

File: Python/Lesson_123_design_hashset.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 123 -- Design HashSet
# Category   : Arrays and Hashing
# Difficulty : Easy
# Study Plan : Day 62
# =============================================================
#
# QUESTION:
#   Design a HashSet (without built-in set): add, remove, contains.
# =============================================================
class MyHashSet:
    def __init__(self): self.b=[[] for _ in range(769)]
    def _h(self,k): return k%769
    def add(self,k):
        b=self.b[self._h(k)]
        if k not in b: b.append(k)
    def remove(self,k):
        b=self.b[self._h(k)]
        if k in b: b.remove(k)
    def contains(self,k): return k in self.b[self._h(k)]

if __name__=="__main__":
    s=MyHashSet(); s.add(1); s.add(2); print(s.contains(1), s.contains(3)); s.remove(2); print(s.contains(2))
